isthmus/src/main/java/io/substrait/isthmus/expression/FunctionMappings.java

```diff
@@ -41,6 +46,24 @@ public class FunctionMappings {
                 s(SqlStdOperatorTable.COUNT, "count"),
                 s(SqlStdOperatorTable.AVG, "avg"))
             .build();
+
```

Comment:
Rather than having two separate sets, maybe we can create an object that contains bidirectional mappings and then just have the object referenced bidirectionally? Feels like that would avoid double-declaring things.

Reply:
The existing mapping of `SCALAR_SIGS` is SqlOperator -> a **category of substrait functions**. 

```
s(SqlStdOperatorTable.PLUS, "add")
s(SqlStdOperatorTable.DATETIME_PLUS, "add"
```

For instance, "add" will expand into add:i64_i64, add:i32_i32, .... in `FunctionConverter.signatures`.  When convert from calcite SqlOperator to substrait function,  we could use argument types to find the matched substrait function. 

When convert from substrait function to SqlOperator, using the above mapping,  a add:i64_i64 would map to `SqlStdOperatorTable.PLUS` and `SqlStdOperatorTable.DATETIME_PLUS`.   The new added `SCALAR_RESOLVER` is to provide a way to different based on the output type.   Calcite code seems also switch to `DATETIME_PLUS` based on output type [1].

If we want to avoid this two separate sets, we need have mapping from SqlOperator <-->. **individual substrait function** in the form of  `add:i64_i64_i64`, rather than a map to a category of functions.

1. https://github.com/apache/calcite/blob/main/core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java#L1040-L1044

Reply:
Yes, I'll think about it. I tried to use the new added way to do a return-type based verification, to choose the calcite SqlOperator.  But a datetime + operator's return type inference would not work for numeric +'s operands binding, that approach does not work. So, the checking is only applied `after` we pick up the operator, but could not use `before` among a set of candidate sql operators.